### DISCUSSION TITLE
Reading data from Spaces

### DISCUSSION THREAD

**Discussion by apeksha-r-naik** [finos/TimeBase-CE#259] [Q&A] [answered]
Hello Team,
Currently we are reading the data from the specific spaces like below

    options = tbapi.SelectionOptions()
    options.space = "20230602"
    cursor = tb_connection.executeQuery(qqlQuery, options)   

But in the result instead of specified space some different space data is getting loaded.

when we run the below query:
select count{} () as TIMEBASE_COUNT, symbol, trade_type, run_date, venue from <table> where symbol=='EURUSD'

We get the below output:
![image](https://github.com/finos/TimeBase-CE/assets/148979235/20e2a440-b31e-4d03-bcd4-a5ac4e85769c)

@alex-karpovich  It is giving us the result from different space using python launguage.

**Comment by ankitsoin** [comment]
Hi @alex-karpovich ,

Using timebase client 6.0.4
Can you please let us know, below is the correct way to read data from spaces of the stream using python?

options = tbapi.SelectionOptions()
options.space = "20230602"
cursor = tb_connection.executeQuery(qqlQuery, options)

We found out that while querying data for space "20230602" timebase is scanning whole stream instead of that particular space.
Or we need to add something else as well for reading from spaces using python?

**Comment by alex-karpovich** [comment]
Greetings, @ankitsoin 
Sorry for long delay, I would recommend to switch to the latest version: **6.1.20** for the TimeBase Server and Client
API changed a bit ti support several spaces in selection (JAVA):
```
SelectionOptions so = new SelectionOptions();
so.withSpace("20230602");
```

**Reply by ankitsoin** [reply]
No Worries, @alex-karpovich

Actually we have a use case where we are using python for fetching aggregated counts of symbol from the spaces of the stream.
So we tried as per documentation as mentioned by @apeksha-r-naik.

We are using Python Timebase client 6.0.4 and server is 6.1.20.
So instead of counts from specific space (e.g "20230602") it is giving counts from whole stream.

Seems spaces is not working in python while fetching data? Can you please suggest is there any work around for same.

**Reply by alex-karpovich** [reply]
We will fix python API to add ability to define list of spaces to read.
@ankitsoin 
Tomorrow next release will be available.

**Reply by ankitsoin** [reply]
Thanks a lot @alex-karpovich for your quick action

**Reply by apeksha-r-naik** [reply]
Thanks @alex-karpovich for you quick response!

We tested the functionality again [reading the data from the spaces using (PYTHON) ] this morning by upgrading TB client version to the latest one i.e. [6.0.5](https://github.com/epam/TimeBaseClientPython/releases/tag/6.0.5), we are still facing the issue, 

Instead of specific space counts (e.g "20230602") it is giving the counts from the whole stream.

Could you please let us know by when we can expect this fix.

**Reply by alex-karpovich** [reply]
Please use TimeBase Server 6.1.21 https://github.com/finos/TimeBase-CE/releases/tag/6.1.21
And use Python API 6.0.5

**Reply by ankitsoin** [reply]
Hi @alex-karpovich,
We have upgraded TB server as well as python client to the above latest version.

But still we are unable to fetch records from particular spaces. We are getting records from all whole stream.

May I know is this code we should use?

options = tbapi.SelectionOptions()
options.space = "20230602"
cursor = tb_connection.executeQuery(qqlQuery, options)

**Reply by Romkisel** [reply, marked as answer]
Hi, @ankitsoin 
Please, try the following api: `options.withSpaces(['Space1', 'Space2', ...])`. For example:

```
            options = tbapi.SelectionOptions()
            options.withSpaces(['20230602'])
            cursor = tb_connection.executeQuery(qqlQuery, options)
```

**Reply by apeksha-r-naik** [reply]
Thank you @Romkisel @alex-karpovich, we are successfully able to read the data from particular spaces using python script.
